server/rotate/rotate_server.go

```diff
@@ -0,0 +1,123 @@
+package rotate
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/http"
+	"strings"
+
+	cjson "github.com/gibson042/canonicaljson-go"
+	"github.com/gorilla/mux"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/fn"
+	"github.com/pingcap/tiup/pkg/cliutil/progress"
+	"github.com/pingcap/tiup/pkg/repository/v1manifest"
+)
+
+type statusRender struct {
+	mbar *progress.MultiBar
+	bars map[string]*progress.MultiBarItem
+}
+
+func newStatusRender(manifest *v1manifest.Manifest, addr string) *statusRender {
+	ss := strings.Split(addr, ":")
+	if strings.Trim(ss[0], " ") == "" || strings.Trim(ss[0], " ") == "0.0.0.0" {
+		addrs, _ := net.InterfaceAddrs()
+		for _, addr := range addrs {
+			if ip, ok := addr.(*net.IPNet); ok && !ip.IP.IsLoopback() && ip.IP.To4() != nil {
+				ss[0] = ip.IP.To4().String()
+				break
+			}
+		}
+	}
+
+	status := &statusRender{
+		mbar: progress.NewMultiBar(fmt.Sprintf("Waiting all administrators to sign http://%s/rotate/root.json", strings.Join(ss, ":"))),
+		bars: make(map[string]*progress.MultiBarItem),
+	}
+	root := manifest.Signed.(*v1manifest.Root)
+	for key := range root.Roles[v1manifest.ManifestTypeRoot].Keys {
+		status.bars[key] = status.mbar.AddBar(fmt.Sprintf("  - Waiting key %s", key))
+	}
+	status.mbar.StartRenderLoop()
+	return status
+}
+
+func (s *statusRender) render(manifest *v1manifest.Manifest) {
+	for _, sig := range manifest.Signatures {
+		s.bars[sig.KeyID].UpdateDisplay(&progress.DisplayProps{
+			Prefix: fmt.Sprintf("  - Waiting key %s", sig.KeyID),
+			Mode:   progress.ModeDone,
+		})
+	}
+}
+
+func (s *statusRender) stop() {
+	s.mbar.StopRenderLoop()
+}
+
+// Serve starts a temp server for receiving signatures from administrators
+func Serve(addr string, root *v1manifest.Root) (*v1manifest.Manifest, error) {
+	payload, err := cjson.Marshal(root)
+	if err != nil {
+		return nil, fn.ErrorWithStatusCode(errors.Annotate(err, "marshal root manifest"), http.StatusInternalServerError)
+	}
+
+	r := mux.NewRouter()
+
+	r.Handle("/rotate/root.json", fn.Wrap(func() (*v1manifest.Manifest, error) {
+		return &v1manifest.Manifest{Signed: root}, nil
+	})).Methods("GET")
+
+	sigCh := make(chan v1manifest.Signature)
+	r.Handle("/rotate/root.json", fn.Wrap(func(m *v1manifest.RawManifest) (*v1manifest.Manifest /* always nil */, error) {
+		for _, sig := range m.Signatures {
+			sigCh <- sig
+		}
+		return nil, nil
+	})).Methods("POST")
+
+	srv := &http.Server{Addr: addr, Handler: r}
+	go func() {
+		err = srv.ListenAndServe()
+		if err == http.ErrServerClosed {
+			err = nil
+		}
+		close(sigCh)
+	}()
+
+	manifest := &v1manifest.Manifest{Signed: root}
+	status := newStatusRender(manifest, addr)
+	defer status.stop()
+
+SIGLOOP:
+	for sig := range sigCh {
+		k := root.Roles[v1manifest.ManifestTypeRoot].Keys[sig.KeyID]
+		if k == nil {
+			// Received a signature signed by an invalid key
+			continue
```

Comment:
Is there any message returned to client to let them know they used a wrong key?

Reply:
Added